odl/discr/discr_ops.py

```diff
@@ -101,56 +111,353 @@ def _call(self, x, out=None):
+            at most as large as the original array shape.
+
+            ``'order0'``: Extend constantly with the outmost values
+            (ensures continuity).
+
+            ``'order1'``: Extend with constant slope (ensures continuity of
+            the first derivative). This requires at least 2 values along
+            each axis where padding is applied.
+
+        pad_const : scalar, optional
+            Value to be used in the ``'constant'`` padding mode.
+
+        discr_kwargs: dict, optional
+            Keyword arguments passed to the `uniform_discr` constructor.
+        """
+        if not isinstance(domain, DiscreteLp):
+            raise TypeError('`domain` must be a `DiscreteLp` instance, '
+                            'got {!r}'.format(domain))
+
+        if not domain.is_uniform:
+            raise ValueError('`domain` is not uniformly discretized')
+
+        offset = kwargs.pop('offset', None)
+        discr_kwargs = kwargs.pop('discr_kwargs', {})
+
+        if range is None:
+            if ran_shp is None:
+                raise ValueError('either `range` or `ran_shp` must be '
+                                 'given')
+
+            offset = normalized_scalar_param_list(
+                offset, domain.ndim, param_conv=safe_int_conv)
+            self.__offset = tuple(offset)
+
+            range = _resize_discr(domain, ran_shp, offset, discr_kwargs)
+
+        elif ran_shp is None:
+            if offset is not None:
+                raise ValueError('`offset` can only be combined with '
+                                 '`ran_shp`')
+
+            if not np.allclose(range.cell_sides, domain.cell_sides):
+                raise ValueError(
+                    'cell sides of domain and range differ significantly '
+                    '(difference {})'
+                    ''.format(range.cell_sides - domain.cell_sides))
+
+            self.__offset = self._offset_from_spaces(domain, range)
+
+        else:
+            raise ValueError('cannot combine `range` with `ran_shape`')
+
+        pad_mode = kwargs.pop('pad_mode', 'constant')
+        self.__pad_mode = str(pad_mode).lower()
+        if self.pad_mode not in _SUPPORTED_RESIZE_PAD_MODES:
+            raise ValueError("`pad_mode` '{}' not understood".format(pad_mode))
+
+        self.__pad_const = float(kwargs.pop('pad_const', 0.0))
+
+        # padding mode 'constant' with `pad_const != 0` is not linear
+        linear = (self.pad_mode != 'constant' or self.pad_const == 0.0)
+
+        super().__init__(domain, range, linear=linear)
+
+    @property
+    def offset(self):
+        """Number of cells added to/removed from the left."""
+        return self.__offset
+
+    @property
+    def pad_mode(self):
+        """Padding mode used by this operator."""
+        return self.__pad_mode
+
+    @property
+    def pad_const(self):
+        """Constant used by this operator in case of constant padding."""
+        return self.__pad_const
+
+    @staticmethod
```

Comment:
If it's useful on its own right, it can be a free function, sure. Here it is a bit more safe to make assumptions on the parameters.